src/libbson/src/bson/bson-utf8.c

```diff
@@ -264,77 +376,104 @@ bson_utf8_escape_for_json (const char *utf8, /* IN */
    bson_unichar_t c;
    bson_string_t *str;
    bool length_provided = true;
+   ssize_t pos;
    const char *end;
 
    BSON_ASSERT (utf8);
 
-   str = bson_string_new (NULL);
-
    if (utf8_len < 0) {
       length_provided = false;
       utf8_len = strlen (utf8);
    }
 
+   if (utf8_len == 0) {
+      return bson_strdup ("");
+   }
+
    end = utf8 + utf8_len;
 
-   while (utf8 < end) {
+   str = bson_string_alloc (utf8_len);
+
+   pos = 0;
+
+   do {
       uint8_t mask;
       uint8_t length_of_char;
 
+      c = (unsigned char) utf8[pos];
```

Comment:
```suggestion
const uint8_t current_byte = (uint8_t) utf8[pos];
```
A separate variable of type `uint8_t` may be required to distinguish it from the UTF-8 character `bson_unichar_t c` when used for single-byte operations. `c` as a `bson_unichar_t` isn't used until the `// Unicode character` block below by the call to `bson_utf8_get_char`.